pss/pss.go

```diff
@@ -553,20 +555,58 @@ func (p *Pss) isSelfPossibleRecipient(msg *PssMsg, prox bool) bool {
 // SECTION: Message sending
 /////////////////////////////////////////////////////////////////////
 
-func (p *Pss) enqueue(msg *PssMsg) error {
+// Enqueue a message msg. Pending will be true for messages with a booked slot in the queue
+func (p *Pss) enqueue(msg *PssMsg, pending bool) error {
 	defer metrics.GetOrRegisterResettingTimer("pss.enqueue", nil).UpdateSince(time.Now())
 
 	outboxmsg := newOutboxMsg(msg)
-	select {
-	case p.outbox <- outboxmsg:
-		metrics.GetOrRegisterGauge("pss.outbox.len", nil).Update(int64(len(p.outbox)))
-
-		return nil
-	default:
+	// We mutex between the calculation of the outbox capacity and the introduction in the channel so we make sure nobody
+	// add a message to the queue after checking outboxCapacity
+	mux := sync.Mutex{}
```

Comment:
You are right, we will study your solution because in that case this mutex won't be needed anyway.

Reply:
Could you explain why the process channel has no buffer? Should not be make(chan int, queueCapacity). If not the enqueue thread will block even when there are empty slots if there are less workers processing the message than messages currently in the outbox.

Reply:
> this mutex is a local variable, totally inert :)

Fixed in commit 462897d.

Reply:
> Could you explain why the process channel has no buffer? Should not be make(chan int, queueCapacity). If not the enqueue thread will block even when there are empty slots if there are less workers processing the message than messages currently in the outbox.

the process channel needs no buffer, the part that sendscan be in

Reply:
@zelig I don't understand, how does this not stop between the range on `process` and reinsert in the channel, no matter how async it is? We want to do the "send and forward" in parallell.

```
// each  worker can just reinsert  async
for i := range process {
    // send and forward
    // if fails to send then 
       select {
          case process <- i:
          case <- quit:
          }
    // if success give back the slot                      
   slots <- i // never blocks
} 
 
```

Reply:
I understood that the for loop is executed for each worker, so you need to spawn as many go routines as workers you want. So the number of workers will be fixed and constant. 
My problem with that piece of code is that a deadlock is possible in a situation where all messages sent fail at any point in time, because the workers are both reading and writing in the bufferless process channel.

Reply:
I'm ok with pursuing the method you originally proposed. But of course without the _locally_ defined mutex

Reply:
Yes the locally declared mutex has already gone.
I must say I like @zelig solution because it uses mutex-less (so more "go way", I guess).
Maybe we can merge the original solution, and then try this other one.

Reply:
Well it doesn't help if the process channel ultimately is blocking anyway :) feel free to pursue that course too, but in this case make sure that we can still do parallell send and retries even though we have intermittent send problems.

Reply:
I created a gist to showcase the deadlock problem. It sovles with a process buffer = outbox size:
https://gist.github.com/kortatu/49093aa5f503a09780314f5a507f12a2

Reply:
That seems reasonable